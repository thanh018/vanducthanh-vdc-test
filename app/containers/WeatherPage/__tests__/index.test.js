import React from 'react';
import { mountWithIntl } from 'tests/intlHelper/intlHelper';
import toJson from 'enzyme-to-json';
import {
  WeatherPage,
  mapDisPatchToProps,
  onGoToDetail,
  weathersToDisplay,
} from '../index';
import {
  weatherDetailMockData,
  locationListMockData,
  weatherDetailMapMockData,
} from './__mocks__/index';

describe('WeatherPage', () => {
  let renderedPage;
  let mockProps;

  const setState = jest.fn();
  const useStateSpy = jest.spyOn(React, 'useState');
  useStateSpy.mockImplementation(init => [init, setState]);

  beforeEach(() => {
    const resetWeatherDetailMock = jest.spyOn(
      mapDisPatchToProps,
      'resetWeatherDetail',
    );
    const fetchSearchDataMock = jest.spyOn(
      mapDisPatchToProps,
      'fetchSearchData',
    );
    const getWeatherDetailMock = jest.spyOn(
      mapDisPatchToProps,
      'getWeatherDetail',
    );

    mockProps = {
      locationList: [],
      weatherDetail: [],
      locationTitle: '',
      isLoading: false,
      resetWeatherDetail: resetWeatherDetailMock,
      fetchSearchData: fetchSearchDataMock,
      getWeatherDetail: getWeatherDetailMock,
    };

    renderedPage = mountWithIntl(<WeatherPage {...mockProps} />, 'en', true);
  });

  afterEach(() => {
    jest.clearAllMocks();
    renderedPage.unmount();
    mockProps = null;
  });

  it('should render renderedPage', () => {
    expect(renderedPage.length).toBe(1);
  });

  it('should match to snapshot', () => {
    expect(
      toJson(renderedPage, {
        mode: 'deep',
        noKey: true,
      }),
    ).toMatchSnapshot();
  });

  describe('Check render results', () => {
    it('should render list results', async () => {
      const event = {
        target: { value: 'm' },
      };
      const searchInput = renderedPage.find(
        'input[data-testid="search-input-location"]',
      );

      searchInput.simulate('change', event);
      renderedPage.setProps({ isLoading: false });
      renderedPage.setProps({ query: 'm' });
      renderedPage.setProps({ locationList: locationListMockData });
      renderedPage.setProps({ weatherDetail: [] });
      const spinComp = renderedPage.find('.ant-spin-nested-loading');
      const listResult = await renderedPage.find(
        '[data-testid="list-location"]',
      );
      expect(spinComp.length).not.toBe(0);
      expect(listResult.length).not.toBe(0);
    });

    it('should call getWeatherDetail when click item of location', async () => {
      const event = {
        target: { value: 's' },
      };
      const searchInput = renderedPage.find(
        'input[data-testid="search-input-location"]',
      );

      searchInput.simulate('change', event);
      renderedPage.setProps({ isLoading: false });
      renderedPage.setProps({ query: 's' });
      renderedPage.setProps({ locationList: locationListMockData });
      renderedPage.setProps({ weatherDetail: [] });
      const itemLocation = await renderedPage.find('.ant-list-item').first(0);

      itemLocation.simulate('click');
      expect(mockProps.getWeatherDetail).toHaveBeenCalledTimes(1);
    });
  });

  describe('Life cycle of components', () => {
    describe('Check fetchSearchData function in useEffect', () => {
      it('should call not fetchSearchData when query state changed before 3s', () => {
        const queryDataMock = 'san';
        renderedPage.setProps({ query: queryDataMock });
        expect(mockProps.fetchSearchData).toHaveBeenCalledTimes(0);
      });

      it('should call fetchSearchData 1 time when query state changed and after 3s', () => {
        const queryDataMock = 'san';
        renderedPage.setProps({ query: queryDataMock });
        setTimeout(() => {
          expect(mockProps.fetchSearchData).toHaveBeenCalledTimes(1);
        }, 300);
      });

      it('should call resetWeatherDetail func when filled search input', () => {
        renderedPage.setProps({ weatherDetail: [{ id: 1 }] });
        const event = {
          target: { value: 'sa' },
        };
        const searchInput = renderedPage.find(
          'input[data-testid="search-input-location"]',
        );

        searchInput.simulate('change', event);
        expect(mockProps.resetWeatherDetail).toHaveBeenCalledTimes(1);
      });

      it('should call not resetWeatherDetail func when weatherDetail empty', () => {
        renderedPage.setProps({ weatherDetail: [] });
        const event = {
          target: { value: 'sa' },
        };
        const searchInput = renderedPage.find(
          'input[data-testid="search-input-location"]',
        );

        searchInput.simulate('change', event);
        expect(mockProps.resetWeatherDetail).toHaveBeenCalledTimes(0);
      });
    });
  });

  describe('Check some function out of side useEffect', () => {
    beforeEach(() => {
      renderedPage.setProps({ query: 'san' });
      renderedPage.setProps({ locationList: locationListMockData });
      renderedPage.setProps({ weatherDetail: [] });
    });

    afterEach(() => {
      renderedPage.setProps({ query: '' });
      renderedPage.setProps({ locationList: [] });
      renderedPage.setProps({ weatherDetail: [] });
    });

    it('should call getWeatherDetail 1 time', () => {
      onGoToDetail(1, mockProps.getWeatherDetail);
      expect(mockProps.getWeatherDetail).toHaveBeenCalledTimes(1);
    });

    it('should call getWeatherDetail 1 time', () => {
      const weathersMap = weathersToDisplay(weatherDetailMockData);
      expect(weathersMap).toEqual(weatherDetailMapMockData);
    });
  });
});
