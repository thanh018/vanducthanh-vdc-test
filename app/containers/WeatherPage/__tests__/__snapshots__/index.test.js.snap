// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`WeatherPage should match to snapshot 1`] = `
<div>
  <h2
    aria-label={null}
    className="ant-typography"
    style={
      Object {
        "WebkitLineClamp": null,
      }
    }
  >
    Weather page
  </h2>
  <div
    className="ant-row"
    style={Object {}}
  >
    <div
      className="ant-col ant-col-6"
      style={Object {}}
    />
    <div
      className="ant-col ant-col-12"
      style={Object {}}
    >
      <div>
        <div
          className="function search() { [native code] } ant-select-show-search ant-select-auto-complete ant-select ant-select-combobox ant-select-enabled"
          onBlur={[Function]}
          onClick={[Function]}
          onFocus={[Function]}
          onMouseDown={[Function]}
          onMouseOut={[Function]}
          onMouseUp={[Function]}
          onTouchStart={[Function]}
          style={
            Object {
              "width": "100%",
            }
          }
        >
          <div
            aria-autocomplete="list"
            aria-controls="test-uuid"
            aria-expanded={false}
            aria-haspopup="true"
            className="ant-select-selection
            ant-select-selection--single"
            role="combobox"
          >
            <div
              className="ant-select-selection__rendered"
            >
              <ul>
                <li
                  className="ant-select-search ant-select-search--inline"
                >
                  <div
                    className="ant-select-search__field__wrap"
                  >
                    <span
                      className="ant-input-search ant-select-search__field ant-input-search-enter-button ant-input-search-large ant-input-group-wrapper ant-input-group-wrapper-lg"
                    >
                      <span
                        className="ant-input-wrapper ant-input-group"
                      >
                        <input
                          className="ant-input ant-input-lg"
                          data-testid="search-input-location"
                          onChange={[Function]}
                          onKeyDown={[Function]}
                          placeholder="Search a location"
                          style={null}
                          type="text"
                          value=""
                        />
                        <span
                          className="ant-input-group-addon"
                        >
                          <button
                            className="ant-btn ant-input-search-button ant-btn-primary ant-btn-lg"
                            onClick={[Function]}
                            type="button"
                          >
                            <i
                              aria-label="icon: search"
                              className="anticon anticon-search"
                            >
                              <svg
                                aria-hidden="true"
                                className=""
                                data-icon="search"
                                fill="currentColor"
                                focusable="false"
                                height="1em"
                                viewBox="64 64 896 896"
                                width="1em"
                              >
                                <path
                                  d="M909.6 854.5L649.9 594.8C690.2 542.7 712 479 712 412c0-80.2-31.3-155.4-87.9-212.1-56.6-56.7-132-87.9-212.1-87.9s-155.5 31.3-212.1 87.9C143.2 256.5 112 331.8 112 412c0 80.1 31.3 155.5 87.9 212.1C256.5 680.8 331.8 712 412 712c67 0 130.6-21.8 182.7-62l259.7 259.6a8.2 8.2 0 0 0 11.6 0l43.6-43.5a8.2 8.2 0 0 0 0-11.6zM570.4 570.4C528 612.7 471.8 636 412 636s-116-23.3-158.4-65.6C211.3 528 188 471.8 188 412s23.3-116.1 65.6-158.4C296 211.3 352.2 188 412 188s116.1 23.2 158.4 65.6S636 352.2 636 412s-23.3 116.1-65.6 158.4z"
                                />
                              </svg>
                            </i>
                          </button>
                        </span>
                      </span>
                    </span>
                    <span
                      className="ant-select-search__field__mirror"
                    >
                      Â 
                    </span>
                  </div>
                </li>
              </ul>
            </div>
            <span
              className="ant-select-arrow"
              onClick={[Function]}
              style={
                Object {
                  "WebkitUserSelect": "none",
                  "userSelect": "none",
                }
              }
              unselectable="on"
            >
              <i
                aria-label="icon: down"
                className="anticon anticon-down ant-select-arrow-icon"
              >
                <svg
                  aria-hidden="true"
                  className=""
                  data-icon="down"
                  fill="currentColor"
                  focusable="false"
                  height="1em"
                  viewBox="64 64 896 896"
                  width="1em"
                >
                  <path
                    d="M884 256h-75c-5.1 0-9.9 2.5-12.9 6.6L512 654.2 227.9 262.6c-3-4.1-7.8-6.6-12.9-6.6h-75c-6.5 0-10.3 7.4-6.5 12.7l352.6 486.1c12.8 17.6 39 17.6 51.7 0l352.6-486.1c3.9-5.3.1-12.7-6.4-12.7z"
                  />
                </svg>
              </i>
            </span>
          </div>
        </div>
      </div>
    </div>
    <div
      className="ant-col ant-col-6"
      style={Object {}}
    />
  </div>
</div>
`;
